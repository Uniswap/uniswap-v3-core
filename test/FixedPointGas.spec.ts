import {MockProvider, deployContract} from 'ethereum-waffle'
import {Contract, BigNumber} from 'ethers'

import FixedPointGasTest from '../build/FixedPointGasTest.json'
import {expect} from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'

const overrides = {
  gasLimit: 9999999,
}

const Q112 = BigNumber.from(2).pow(112)

describe('FixedPoint', () => {
  const provider = new MockProvider({
    ganacheOptions: {
      hardfork: 'istanbul',
      mnemonic: 'horn horn horn horn horn horn horn horn horn horn horn horn',
      gasLimit: 9999999,
    },
  })
  const [wallet] = provider.getWallets()

  let fixedPointGas: Contract
  before('deploy FixedPointExtraTest', async () => {
    fixedPointGas = await deployContract(wallet, FixedPointGasTest, [], overrides)
  })

  describe('#muluq', () => {
    it('short circuits for zero', async () => {
      expect(await fixedPointGas.muluqGasUsed([BigNumber.from(0)], [Q112.mul(22).div(10)])).to.eq('231')
    })

    it('gas', async () => {
      await snapshotGasCost(fixedPointGas.muluqGasUsed([Q112.mul(35).div(10)], [Q112.mul(22).div(10)]))
    })
  })

  describe('#divuq', () => {
    it('gas', async () => {
      await snapshotGasCost(fixedPointGas.divuqGasUsed([Q112.mul(35).div(10)], [Q112.mul(22).div(10)]))
    })
  })
})
