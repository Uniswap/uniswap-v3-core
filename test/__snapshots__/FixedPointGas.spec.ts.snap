// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedPoint #divuq gas 1`] = `461`;

exports[`FixedPoint #muluq gas 1`] = `675`;
